Added a dashboard for local volume provisioner.

chhsia0 opened the pull request:
JIRA: https://jira.mesosphere.com/browse/DCOS-55515

chhsia0 commented on dashboards/localvolumeprovisioner.yaml:
```diff
@@ -0,0 +1,470 @@
+                   "format": "time_series",
+                   "instant": false,
+                   "interval": "",
+                   "intervalFactor": 4,
```
Based on some experiments I did, it looks like that the 1/1 resolution makes the graph a bit noisier, and the 1/2 resolution seems to deliver almost the same amount of information, especially in the heatmap.

Since the heatmap is now removed, the 1/1 resolution looks not as noisy in the new graphs. I'll go with your suggestion.

chhsia0 commented on dashboards/localvolumeprovisioner.yaml:
```diff
@@ -0,0 +1,470 @@
+             "steppedLine": true,
+                "targets": [
+                {
+                   "expr": "idelta(local_volume_provisioner_apiserver_requests_total{instance=~\"$instance\"}[1m])",
```
The metrics are all cumulative, so some kind of difference function is needed to present the state of the provisioner of a given time. We have several other choices here:
`rate`: The end results would be a floating point number that's very hard to understand.
`increase`: This is better than `rate` since it gives the difference within the time interval. However this function will do extrapolation, which would end up with weird floating point numbers that does't match the heatmap.
`delta`: This function also gives similar results to `increase`. But if there's only one single value in a given duration the data will be hidden.

Therefore I ended up with `idelta`, where each point on the graph consumes two data points, and that's why I set the resolution to 1/2. This makes the request graph match the heatmaps and gives the exact number of requests. It also works if there are consecutive nonzero data points.

That said, if we don't need to show the histogram heatmaps, then we don't need this.

chhsia0 commented:
@nfnt I merged you changes and made the right-hand-side graphs showing "null" durations as zeros to make them a bit prettier ;). I didn't change the left-hand-side graphs, so nulls can reflects if a provisioner is malfunctioning.

Overall the dashboard looks good. But I feel the heatmap can still convey some value. Say if there are 100 volumes created at once and only one is slow, it might not be obvious to observe through the average graphs. But we can decide if we want to put the heatmaps back later.

Also how about combining the discovered and deleted volumes together to show the number of "discovered volumes that currently exists ( = discovered - deleted)?" No strong opinion though.

nfnt commented:
I'm opposed to using `idelta`, we should use `increase` or `rate` instead. The problem with `idelta` is that it only takes the last two values of a range-vector for its calculation. While this allows us to get exact values that match the number of operations at that time, it will result in missing or wrong values if even a single metric is missing. This isn't that likely to happen with `rate` as it uses all values of a range-vector. With a 1m window, we would need 3 missing metrics to lose a calculation.

Showing null durations as zero is indeed prettier, but causes an issue when showing the averages: We're dividing two different metrics. This division should only be made if both metrics are available. This isn't the case for `_sum` and `_count` metrics. The result are slopes that can be misinterpreted.
![Null as Null](https://user-images.githubusercontent.com/2085586/61114949-b0817c80-a491-11e9-84ae-875f65334dbc.png)
![Null as Zero](https://user-images.githubusercontent.com/2085586/61114986-c68f3d00-a491-11e9-94b7-c81d26fe0977.png)

chhsia0 commented:
@nfnt Thanks for the feedback! I changed the `idelta` to `increase` (picked it over `rate` because it seems easier to understand that we're doing an average on sum and count, not over time).

I still keep the 30s interval because that's the metric scrape interval defined in prometheus:
https://github.com/mesosphere/kubeaddons-configs/blob/b8d6475eb0405d6ee2282037260b8d32b4474dbb/templates/prometheus.yaml#L48
https://github.com/mesosphere/kubeaddons-configs/blob/b8d6475eb0405d6ee2282037260b8d32b4474dbb/templates/prometheus.yaml#L59
Leaving the scrape interval for the Grafana dashboard default would duplicate every data point and make the interpolation misleading.

The missing data points is not because `_count` is non-null but `_sum` is. Prometheus's histogram metrics should always generate both: https://prometheus.io/docs/concepts/metric_types/#histogram
The problem is that when there's no incoming requests, both `_sum` and `_count` would be 0, so 0 / 0 = NaN, which is not rendered in the graph.

I don't think the slope is that misleading. But to address this, I've changed it to use stepped lines.
